### ISSUE TITLE
Add Thai keyboard

### ISSUE THREAD

**Issue by Yyatiko** [koreader/koreader#5615]
* KOReader version:  v2019.08
* Device: Kobo_pika

#### Issue

Hi,

I installed a Thai Dictionary on my e-reader and I would like to enter queries in Thai Language in Koreader.

Here is a link of two typical Thai dictionaries : https://www.branah.com/thai

I see that I can add Thai characer in the [en_popup.lua files](https://github.com/koreader/koreader/tree/48f20b215350576133abbf85b7f0e94f99046d08/frontend/ui/data/keyboardlayouts/keypopup), but it would be more convenient to be able to shift the entire keyboard to a thai keyboard. 

Also I was wondering how to change the default keyboard shift that shows up when we press the planet character at the bottom left of the english keyboard.

Is there a way to automatically create a .lua files form an arbitrary keyboard file ?
Also I could try to create my own Thai keyboard in a .lua file similar to en_keyboard.lua. But how to install it afterward ?

Thank you for your help.

**Comment by Frenzie** [comment]
The keyboards were actually completely overhauled yesterday in #5583 and #5610.

If you update to the latest development version you can set enabled keyboards in Device → Keyboard layout for quick switching.

> Is there a way to automatically create a .lua files form an arbitrary keyboard file ?
Also I could try to create my own Thai keyboard in a .lua file similar to en_keyboard.lua. But how to install it afterward ?

You would replace the QWERTY characters with the Kedmanee layout. You can add more rows if desired, see the Japanese layout.

To make it show up, add it to the list here:

https://github.com/koreader/koreader/blob/15fe0fef07413b54f847fdffbc601f05d78fb3a7/frontend/ui/widget/virtualkeyboard.lua#L651-L661

**Comment by Yyatiko** [comment]
Ok great thank you.

I tried the version from [Novembre 18](http://build.koreader.rocks/download/nightly/v2019.11-20-g15fe0fe_2019-11-18/) but it crashes when I select a languages from the pop up menu (sorry for poor quality)

![Photo le 19-11-19 à 13 42 #2](https://user-images.githubusercontent.com/54907189/69175878-9a846800-0ad2-11ea-8472-839fedebdd00.jpg)

**Comment by Frenzie** [comment]
That's correct, I fixed that a couple of hours ago in #5616. But it will nevertheless work without that patch if you select a few in Settings → Device → Keyboard layout.

**Comment by Yyatiko** [comment]
Very nice, you are right, it works.

So I just need to built a keyboard file now.
Thank you very much.

**Comment by Yyatiko** [comment]
Hi,

So I tried to configure a new keyboardlayout as mention modifying the Japanese keyboard.

The only thing is that some of the vowels in Thai are like accent and get added on top or bellow or in front or in the back of a consonant

see this keyboard for try https://www.branah.com/thai 

My modify Japanese layout is like this so far:

{  --  1       2       3       4
            { "_",    "%",    "~",    "`", },
            { "ๅ",    "+",    "!",    "1", },
            { "/",    "๑",    "@",    "2", },
            { "-",    "๒",    "#",    "3", },
            { "ภ",    "๓",    "$",    "4", },
            { " ุ",    "๔",    "%",    "5", },
            { " ึ",    " ู",    "^",    "6", },
            { "ต",    "฿",    "&",    "7", },
            { "ต",    "๕",    "*",    "8", },
            { "จ",    "๖",    "(",    "9", },
            { "ข",    "๗",    ")",    "0", },
            { "ช",    "๘",    "_",    "-", },
            { "゛",    "๙",    "+",    "=", },

You will notice the vowels in the middle with the little circle to describe where the vowel attaches on the consonant. Unfortunately it does't work on the kobo reader using a simple copy paste for those vowels as key from https://www.branah.com/thai 

Any suggestions ?
Thanks

**Comment by poire-z** [comment]
> Unfortunately it does't work on the kobo reader 

You need to tell us more what you mean by "it doesn't work" :)
Does it crash?
Or is it just that the letters are outputed, but are not combined/merged/shaped as you expect?

In the case Thai needs some complex characters/glyphs shaping (and it looks like [Harfbuzz has some specific for it](https://github.com/harfbuzz/harfbuzz/blob/master/src/hb-ot-shape-complex-thai.cc)), you might need to update to [a recent nightly](http://build.koreader.rocks/download/nightly/v2019.11-51-ge8f91ab_2019-12-03/). There's been some significant improvement on text shaping since 2019.11, so I'm quite intrigued to see if all that we did for Arabic also makes Thai magically looks allright!

About Thai:
https://en.wikipedia.org/wiki/Thai_(Unicode_block)
http://www.alanwood.net/unicode/thai.html
https://r12a.github.io/scripts/thai/

**Comment by Yyatiko** [comment]
Yes, when I say it doesn't work, I mean that the vowel don't display at the right spot on the consonant or is even a different symbol...

I will try to upgrade soon and make another test.

Thank you for you answer.

**Comment by poire-z** [comment]
I think it will work:
<kbd>![image](https://user-images.githubusercontent.com/24273478/70162369-d405c780-16bd-11ea-8c48-8bcbc216502f.png)</kbd>

But you'll have to think a little bit more about how you want your keyboard.

Your vowels seem to be "diacritics", like my french accents: some exist in Unicode standalone, and they are put on the real character before it - or they exist combined as a single character. Dunno what's available for Thai (but see https://r12a.github.io/scripts/thai/block), and how you are used to type things, 2 keys, or single keys for each combination.

In french: <kbd>a</kbd> + <kbd>^</kbd> gives <kbd>â</kbd>, as we don't have a key for <kbd>â</kbd> directly - but we have one for <kbd>à</kbd>, that we can also type as <kbd>a</kbd>+<kbd>`</kbd>.

I used U+0E37 THAI CHARACTER SARA UEE for the test above. On the first key, it's combined with one of your consonant (I guess). On the 2nd key, it's a standalone (you can see it's in a little circle, but it is off the key (and bleeds on the preceeding key :| ) for a reason I don't know, but we'll probably find a way to fix it).
On the keys you copied, the diacritic in the middle does not seem to be a diacritic, as it doesn't combine.
(Or I dunno, may be they are more than simple diacritics, and characters on their own that really should shape with their neighbour glyphs when needed.)

Anyway, you might want to look at the hebrew keyboard that uses a lot of diacritics:
<kbd>![image](https://user-images.githubusercontent.com/24273478/70162077-5fcb2400-16bd-11ea-953f-288160fe7c9d.png)</kbd>
and it also has popups, so, with a long press on a consonnant, you could have a popup showing up to 11 combinations of it with a vowel, so your Thai user has the choice of combining or just using pre-combined forms:
<kbd>![image](https://user-images.githubusercontent.com/24273478/70162258-a6b91980-16bd-11ea-874d-8516abc3ad21.png)</kbd>

You can also have a look at our discussions while the Hebrew #5414 and Arabic #5569 were being made.
